_posts/2018-10-23-grpc-web-ga.md

```diff
@@ -0,0 +1,122 @@
+
+<img src="/img/grpc-web-arch.png" style="max-width: 947px">
+
+
+<p style="text-align: center"> Figure 1.
+gRPC with gRPC-Web (left) and gRPC with REST (right)</p>
+
+In the gRPC-Web universe on the left, a client application speaks Protocol Buffers to a gRPC backend server that speaks Protocol Buffers to other gRPC backend services. In the REST universe on the right, the web app speaks HTTP to a backend REST API server that then speaks Protocol Buffers to backend services.
+
+## Advantages of using gRPC-Web
+gRPC-Web will offer an ever-broader feature set over time, but here’s what’s in 1.0 today:
+
+* **End-to-end gRPC** — Enables you to craft your entire RPC pipeline using Protocol Buffers. Imagine a scenario in which a client request goes to an HTTP server, which then interacts with 5 backend gRPC services. There’s a good chance that you’ll spend as much time building the HTTP interaction layer as you will building the entire rest of the pipeline. 
+* **Tighter coordination between frontend and backend teams**  — With the entire RPC pipeline defined using Protocol Buffers, you no longer need to have your “microservices teams” alongside your “client team.” The client-backend interaction is just one more gRPC layer amongst others. 
+* **Easily generate client libraries**  — With gRPC-Web, the server that interacts with the “outside” world, i.e. the membrane connecting your backend stack to the internet, is now a gRPC server instead of an HTTP server, that means that all of your service’s client libraries can be gRPC libraries. Need client libraries for Ruby, Python, Java, and 4 other languages? You no longer need to write HTTP clients for all of them.
+
+## A gRPC-Web example
+The previous section illustrated some of the high-level advantages of gRPC-Web for large-scale applications. Now let’s get closer to the metal with an example: a simple TODO app. In gRPC-Web you can start with a simple ``todos.proto`` definition like this:
+```proto
+syntax = "proto3";
+
+package todos;
+
+message Todo {
+  string content = 1;
+  bool finished = 2;
+}
+
+message GetTodoRequest {
+  int32 id = 1;
+}
+
+service TodoService {
+  rpc GetTodoById (GetTodoRequest) returns (Todo);
+} 
+
+```
+CommonJS client-side code can be generated from this ``.proto`` definition with the following command:
+
+```proto
+protoc echo.proto \
+  --js_out=import_style=commonjs:./output \
+  --grpc-web_out=import_style=commonjs:./output
+```
+
+Now, fetching a list of TODOs from a backend gRPC service is as simple as:
+
+```js
+const {GetTodoRequest} = require('./todos_pb.js');
+const {TodoServiceClient} = require('./todos_grpc_web_pb.js');
+
+const todoService = new proto.todos.TodoServiceClient('http://localhost:8080');
+const todoId = 1234;
+
+var getTodoRequest = new proto.todos.GetTodoRequest();
+getTodoRequest.setId(todoId);
+
+var metadata = {};
+var getTodo = todoService.getTodoById(getTodoRequest, metadata, (err, response) => {
+  if (err) {
+    console.log(err);
```

Comment:
not `console.error`?

Reply:
Convention is usually just `console.log` for everything.